### DISCUSSION TITLE
Airflow Kubernetes Executor Pods Not launching due to permissions

### DISCUSSION THREAD

**Discussion by kl4072** [apache/airflow#37574] [Q&A] [answered]
In current environment cannot create new Role or rolebindings
Role has been created and bound by the admin so i have disabled the creation of pod-launcher-role, rolebinding .. does the Role need to list resources in default namespace? why is it trying to list in default namespace even though the namespace is explicitly defined in pod-template?

HTTP response body: b'{\n "kind": "Status",\n "apiVersion": "v1",\n "metadata": {},\n "status": "Failure",\n "message": "pods is forbidden: User \"system:serviceaccount:orchestration-dev:airflow-service-account\" cannot list resource \"pods\" in API group \"\" in the namespace \"default\"",\n "reason": "Forbidden",\n "details": {\n "kind": "pods"\n },\n "code": 403\n}'

**Comment by hanleybrand** [comment]
Are you running airflow in the default namespace?  

In my groups cluster (on an institutional k8s server) by default the cluster owner doesn't have as many permissions in the default namespace, because it's not "supposed" to be used (the [official docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#when-to-use-multiple-namespaces) say "For a production cluster, consider not using the default namespace. Instead, make other namespaces and use those" but I've seen other things that say default should not be used, and I think this has permeated into the realm of best practices among k8s cluster admins.

If you aren't running airflow in the default namespace (e.g. you're using `namespace: airflow`), and assuming you're talking about dags launching new pods using the base image (i.e. you aren't programmatically creating pods, you're just writing dags and airflow is handling the pods) here's some suggestions/things to check going by the 1.12 version of the helm chart values.yaml file:

- I'm pretty sure you should have `multiNamespaceMode: false` (keeps pods in airflow's namespace)
- check the airflow configmap as deployed with kubectl, etc to make sure that the [kubernetes] & [kubernetes_executor] have agreeing and correct values and also contain `multi_namespace_mode = False`, `namespace = airflow`, &c. i.e. that the right values are in the config
- it's commented out in the 1.12 chart's values.yaml file, but it looks like you can set env.AIRFLOW_VAR_KUBERNETES_NAMESPACE  in the config, although I'm pretty sure this is new to the 1.12 chart
- Also, make sure `executor: KubernetesExecutor` is a root value in your values.yaml file (see [this issue](https://github.com/apache/airflow/issues/29365#issuecomment-1435988327) and not indented)

**Reply by kl4072** [reply]
namespace -orchestration-dev is where all the containers are currently running lets say the account is a local cluster account with Cluster Role privilege's to this namespace.  this is the settings currently
 [kubernetes_executor]
multi_namespace_mode = False
namespace = orchestration-dev
pod_template_file = /opt/airflow/pod_templates/pod_template_file.yaml
worker_container_repository = apache/airflow
worker_container_tag = 2.8.1-python3.11

executor: KubernetesExecutor is set on values.yaml

Also have it set at env variable level as follows 
 - name: AIRFLOW__CORE__EXECUTOR
      value: KubernetesExecutor

**Reply by kl4072** [reply]
airflow-service-account is the service account lets say

**Reply by hanleybrand** [reply]
And the scheduler, triggerer and webserver are starting up?  Can you post any other logs or the code of the dag that's not spinning up correctly?

**Reply by hanleybrand** [reply]
One other thought, assuming you're installing the helm chart via the helm cli, maybe re-enable the role creation and run this command, and then compare any generated RoleBinding or ClusterRoleBinding with your cluster configuration?

```bash
helm upgrade airflow apache-airflow/airflow --install -f values.yaml \
-n orchestration-dev --version 1.12.0 --dry-run >> dry_run.txt  2>&1
```

(in my experience, omitting the `2>&1` lead to a very empty dry_run.txt file)

**Reply by kl4072** [reply]
with LocalExecutor, scheduler, webserver all run fine. Using posgresql for backend.
With KubernetesExecutor the webserver runs but scheduler fails due to the above message, only difference with these on the dry run was that the kubernetesExecutor creates a service account. I will run these on my desktop as well to compare

**Reply by kl4072** [reply, marked as answer]
The issue with this was related to us not mounting the airflow.cfg in the scheduler deployment correctly in the helm templates. we had to do quite a bit of customizations on the templates and hence the miss.
